erts/emulator/beam/erl_bif_persistent.c

```diff
@@ -750,6 +664,174 @@ erts_init_persistent_dumping(void)
  * Local functions.
  */
 
+/*
+ * put and put_new helpers.
+ */
+
+static void do_update(ErtsPersistentTermPutContext* ctx)
+{
+    Uint term_size;
+    Uint lit_area_size;
+    ErlOffHeap code_off_heap;
+    ErtsLiteralArea* literal_area;
+    erts_shcopy_t info;
+    Eterm* ptr;
+    /*
+     * Preserve internal sharing in the term by using the
+     * sharing-preserving functions. However, literals must
+     * be copied in case the module holding them are unloaded.
+     */
+    INITIALIZE_SHCOPY(info);
+    info.copy_literals = 1;
+    term_size = copy_shared_calculate(ctx->tuple, &info);
+    ERTS_INIT_OFF_HEAP(&code_off_heap);
+    lit_area_size = ERTS_LITERAL_AREA_ALLOC_SIZE(term_size);
+    literal_area = erts_alloc(ERTS_ALC_T_LITERAL, lit_area_size);
+    ptr = &literal_area->start[0];
+    literal_area->end = ptr + term_size;
+    ctx->tuple = copy_shared_perform(ctx->tuple, term_size, &info, &ptr, &code_off_heap);
+    ASSERT(tuple_val(ctx->tuple) == literal_area->start);
+    literal_area->off_heap = code_off_heap.first;
+    DESTROY_SHCOPY(info);
+    erts_set_literal_tag(&ctx->tuple, literal_area->start, term_size);
+
+    if (ctx->hash_table == (HashTable *) erts_atomic_read_nob(&the_hash_table)) {
+        /* Schedule fast update in active hash table */
+        fast_update_index = ctx->entry_index;
+        fast_update_term = ctx->tuple;
+    }
+    else {
+        /* Do update in copied table */
+        set_bucket(ctx->hash_table, ctx->entry_index, ctx->tuple);
+    }
+
+    /*
+     * Now wait thread progress before making update visible to guarantee
+     * consistent view of table&term without memory barrier in every get/1.
+     */
+    erts_schedule_thr_prgr_later_op(table_updater, ctx->hash_table, &thr_prog_op);
+}
+
+static ErtsPersistentTermPutCommonResult put_common
+(Process* c_p, Eterm key, Eterm term, bool new)
+{
+    static const Uint ITERATIONS_PER_RED = 32;
+    ErtsPersistentTermPutContext* ctx;
+    Eterm state_mref = THE_NON_VALUE;
+    Eterm old_bucket;
+    long iterations_until_trap;
+    long max_iterations;
+#define PUT_TRAP_CODE                                            \
+    BIF_TRAP2(BIF_TRAP_EXPORT(BIF_persistent_term_put_2), c_p, state_mref, term)
+#define TRAPPING_COPY_TABLE_PUT(TABLE_DEST, OLD_TABLE, NEW_SIZE, COPY_TYPE, LOC_NAME) \
```

Comment:
I've managed to pin point the previous trapping errors to here with some printf debugging. `TRAPPING_COPY_TABLE_PUT` stops just as when it runs `PUT_TRAP_CODE`.

I suspect the error is in calling `BIF_TRAP_EXPORT(BIF_persistent_term_put_2)` within `put_common`?